### PR TITLE
SONARSLANG-4 Rule: Unused Function Parameter

### PR DESCRIPTION
**Pull request by marcelo-sousa-sonarsource**

### CODE REVIEW

File: slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+public class UnusedFunctionParameterCheck implements SlangCheck {
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(FunctionDeclarationTree.class, (ctx, functionDeclarationTree) -> {
+      if (!(ctx.parent() instanceof TopLevelTreeImpl)) {
+        return;
+      }
+
+      if (functionDeclarationTree.body() == null || functionDeclarationTree.body().descendants() == null) {
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java]: You could merge this if with the previous one for conciseness. Also `descendants()` cannot be null

File: slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      ParameterTree firstUnused = unusedParameters.get(0);
+      String msg;
+
+      if (unusedParameters.size() > 1) {
+        msg = "Remove these unused function parameters.";
+      } else {
+        msg = "Remove this unused function parameter \"" + firstUnused.identifier().name() + "\".";
+      }
+
+      ctx.reportIssue(firstUnused, msg, secondaryLocations);
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java]: The rule implementation in other analyzers highlight only the identifiers. I think we should do the same here

File: its/ruling/pom.xml [hunk 3]
```diff
@@ -27,5 +27,10 @@
       <artifactId>sonar-analyzer-commons</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.sonarsource.sonar-lits-plugin</groupId>
+      <artifactId>sonar-lits-plugin</artifactId>
```
Comment (christophe-zurn-sonarsource) [inline its/ruling/pom.xml]: Why this new dependency?

File: slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import static com.sonarsource.slang.utils.SyntacticEquivalence.areEquivalent;
+
+@Rule(key = "S1172")
+public class UnusedFunctionParameterCheck implements SlangCheck {
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(FunctionDeclarationTree.class, (ctx, functionDeclarationTree) -> {
+      if (!(ctx.parent() instanceof TopLevelTreeImpl)) {
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java]: That's an example of False Negative we might want to track. Only global functions will raise an issue for now. Class methods and lambda will never raise issue for now. You could create a jira ticket related to this

File: slang-checks/src/test/resources/com/sonarsource/slang/checks/UnusedFunctionParameter.slang [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+void fun fooBar(int a, int b) {}  // Noncompliant {{Remove these unused function parameters.}}
+//              ^^^^^  ^^^^^<
+
+void fun fooBar(int a, int b) {   // Noncompliant {{Remove this unused function parameter "b".}}
+//                     ^^^^^
+  x = a;
+}
+
+void fun fooBar(int a, int b) {   // Noncompliant {{Remove this unused function parameter "a".}}
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/test/resources/com/sonarsource/slang/checks/UnusedFunctionParameter.slang]: We could have one of these test with a parameter without type to ensure it works in that case too

File: slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+          .map(SecondaryLocation::new)
+          .collect(Collectors.toList());
+
+      ParameterTree firstUnused = unusedParameters.get(0);
+      String msg;
+
+      if (unusedParameters.size() > 1) {
+        msg = "Remove these unused function parameters.";
+      } else {
+        msg = "Remove this unused function parameter \"" + firstUnused.identifier().name() + "\".";
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/UnusedFunctionParameterCheck.java]: In java analyzer, they always have this message, and each secondary location also contains this message.
Cf: https://sonarcloud.io/project/issues?id=org.apache.tika%3Atika&issues=8246e5b6-acb7-490e-b1ad-7dd56827218c&open=8246e5b6-acb7-490e-b1ad-7dd56827218c

File: its/ruling/pom.xml (hunk 3 above)
Comment (marcelo-sousa-sonarsource) [reply its/ruling/pom.xml]: @pynicolas suggested we added this dependency to debug ITs locally.

Comment (christophe-zurn-sonarsource) [reply its/ruling/pom.xml]: Ah ok :+1:

Comment (pynicolas) [reply its/ruling/pom.xml]: Orchestrator now looks into the local Maven repository to find the lits plugin. This new dependency forces Maven to download it before running the tests. Potential problem: it has to be the same version as used in the test.

**Pull request merged**